x/clerk/keeper/grpc_query.go

```
@@ -166,6 +166,40 @@ func (q queryServer) IsClerkTxOld(ctx context.Context, request *types.RecordSequ
 	return &types.IsClerkTxOldResponse{IsOld: true}, nil
 }
 
+// GetLatestRecordId implements the gRPC service handler to query the latest record id from L1.
+func (q queryServer) GetLatestRecordId(ctx context.Context, request *types.LatestRecordIdRequest) (*types.LatestRecordIdResponse, error) {
+	if request == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	// Get chain params to get the StateSender contract address.
+	chainParams, err := q.k.ChainKeeper.GetParams(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// Get the StateSender contract instance.
+	stateSenderInstance, err := q.k.contractCaller.GetStateSenderInstance(chainParams.ChainParams.StateSenderAddress)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "failed to get state sender instance")
+	}
+
+	// Get the current state counter from L1.
+	stateCounter := q.k.contractCaller.CurrentStateCounter(stateSenderInstance)
+	if stateCounter == nil {
+		return nil, status.Error(codes.Internal, "failed to get latest state counter from L1")
+	}
+
+	latestRecordId := stateCounter.Uint64()
+
+	_, err = q.k.GetEventRecord(ctx, latestRecordId)
+	if err == nil {
+		return &types.LatestRecordIdResponse{LatestRecordId: latestRecordId, IsProcessedByHeimdall: true}, nil
+	} else {
+		return &types.LatestRecordIdResponse{LatestRecordId: latestRecordId, IsProcessedByHeimdall: false}, nil
```

If you get an error, you should propagate the error instead of returning false by default.
This is because iirc if the event doesn't exist in the collection, you get the 0-value for it (surely not an error), so the info here could be misleading.
I'd say we need to:
- return error if error
- return true if the event record is found
- return false if the event record is not found (zero value)

---

Used `HasEventRecord` for simplicity, nice suggestion.